### PR TITLE
Give all declarations an explicit interface type

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by graydon** [comment]
This is causing failures in LLDB testsuite, eg.
https://ci.swift.org/job/oss-lldb-incremental-osx/4898/

Reverting for now. Sorry, it's way out of my depth to fix.

**Comment by slavapestov** [comment]
No worries, I'll look at it later.

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -2642,6 +2642,7 @@ class NominalTypeDecl : public GenericTypeDecl, public IterableDeclContext {
 protected:
   Type DeclaredTy;
   Type DeclaredTyInContext;
+  Type DeclaredInterfaceTy;
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Why do you need a new variable for this?

File: test/api-digester/source-stability.swift.expected [hunk 2]
```diff
@@ -25,6 +25,14 @@ Var Set.startIndex has declared type change from SetIndex<Element> to Set<Elemen
 Func Dictionary.index(after:) has return type change from DictionaryIndex<Key, Value> to Dictionary<Key, Value>.Index
 Func Dictionary.index(forKey:) has return type change from DictionaryIndex<Key, Value>? to Dictionary<Key, Value>.Index?
 Func Dictionary.remove(at:) has 1st parameter type change from DictionaryIndex<Key, Value> to Dictionary<Key, Value>.Index
+
+/* False positives due to archetype -> interface type change */
```
Comment (jrose-apple) [inline test/api-digester/source-stability.swift.expected]: This file doesn't need to be sorted, so it's probably better to keep the Set/Dictionary stuff together.

**Pull request merged**